goplay/src/main/java/be/tapped/goplay/content/ProgramRepo.kt

```diff
@@ -3,48 +3,37 @@ package be.tapped.goplay.content
 import arrow.core.Either
 import arrow.core.computations.either
 import arrow.core.flatMap
-import arrow.core.left
-import be.tapped.common.internal.executeAsync
 import be.tapped.goplay.ApiResponse.Failure
-import be.tapped.goplay.ApiResponse.Failure.HTML
 import be.tapped.goplay.ApiResponse.Success
+import be.tapped.goplay.common.ktorClient
 import be.tapped.goplay.common.safeAttr
-import be.tapped.goplay.common.safeBodyString
-import be.tapped.goplay.common.safeSelect
 import be.tapped.goplay.common.safeSelectFirst
 import be.tapped.goplay.common.siteUrl
+import io.ktor.client.HttpClient
+import io.ktor.client.request.get
+import io.ktor.client.statement.HttpResponse
+import io.ktor.client.statement.readText
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.withContext
 import kotlinx.serialization.decodeFromString
 import kotlinx.serialization.json.Json
 import kotlinx.serialization.json.JsonObject
 import kotlinx.serialization.json.decodeFromJsonElement
 import kotlinx.serialization.json.jsonObject
-import okhttp3.OkHttpClient
-import okhttp3.Request
-import okhttp3.Response
 import org.jsoup.Jsoup
 import org.jsoup.nodes.Document
 
-internal class HtmlProgramParser(private val jsoupParser: JsoupParser) {
-    internal suspend fun parse(html: String): Either<HTML, List<Program>> = either {
-        val htmlPrograms = jsoupParser.parse(html).safeSelect("a[data-program]").bind()
-        // The Program detail is found within the DOM as a JSON String
-        val jsons = htmlPrograms.map { it.safeAttr("data-program").toEither().bind() }
-        jsons.map {
-            Json {
-                isLenient = true
-                ignoreUnknownKeys = true
-            }.decodeFromString(it)
-        }
-    }
+internal class HtmlProgramParser(private val jsonSerializer: Json = be.tapped.goplay.common.jsonSerializer) {
+    private val regex = "data-program=\"(.*)\"".toRegex()
+    internal fun parse(html: String): Either<Failure, List<Program>> =
+        Either.catch(regex.findAll(html).map { it.groupValues[1] }.map<String, Program>(jsonSerializer::decodeFromString)::toList).mapLeft(Failure::JsonParsingException)
```

Comment:
Actually failing because the extracted JSON is HTML decoded.

`{&quot;id&quot;:&quot;9a0484e7-ca02-4b19-9199-2fbe6c472c12&quot;,&quot;title&quot;:&quot;A Christmas Break&quot;,&quot;link&quot;:&quot;\/a-christmas-break&quot;,&quot;label&quot;:&quot;&quot;,&quot;needs16PlusLabel&quot;:false,&quot;category&quot;:&quot;Film&quot;,&quot;images&quot;:{&quot;poster&quot;:&quot;https:\/\/images.goplay.be\/2021-12\/onlinempl1050x1500a-christmas-break-r4l6cj.jpg?auto=format&amp;dpr=2&amp;fit=clip&amp;h=290&amp;ixlib=php-1.1.0&amp;q=95&amp;w=200&amp;s=d2b698eaff020f701d0abeb717c07c96&quot;,&quot;teaser&quot;:&quot;https:\/\/images.goplay.be\/meta\/onlinewallpaper3840x2160a-christmas-break-r4l6bg.jpg?auto=format&amp;dpr=2&amp;fit=crop&amp;h=max-185&amp;ixlib=php-1.1.0&amp;q=85&amp;w=335&amp;s=830a67dc03829b7e6b3add080e98693a&quot;},&quot;pageInfo&quot;:{&quot;brand&quot;:&quot;Play7&quot;}}"`